Add support for URL-encoded form parameters

mangstadt opened the pull request:
These kinds of form parameters are typically found in the body of POST requests when a form is submitted on a website. Don't know if this would be useful to add or not.

Examples:

```java
client.onPost("/login")
  .withFormParameter("username", "John")
  .withFormParameter("password", Matchers.any(String.class))
.doReturnStatus(200);
```

```java
Map<String, Matcher<String>> parameters = new HashMap<>();
parameters.put("username", Matchers.equalTo("John"));
parameters.put("password", Matchers.any(String.class));
client.onPost("/login")
  .withBody(parameters)
.doReturnStatus(200);
```

```java
client.onGet("/foo").doReturnFormParams(Arrays.asList(
  new BasicNameValuePair("one", "1"),
  new BasicNameValuePair("one", "2")
));
```

mangstadt committed:
Fix syntax errors in code samples

mangstadt committed:
Add support for URL-encoded form parameters

Typically found in the body of POST requests when a form is submitted on
a website.

Examples:

client.onPost("/login")
  .withFormParameter("username", "John")
  .withFormParameter("password", Matchers.any(String.class))
.doReturnStatus(200);

Map<String, Matcher<String>> parameters = new HashMap<>();
parameters.put("username", Matchers.equalTo("John"));
parameters.put("password", Matchers.any(String.class));
client.onPost()("/login")
  .withBody(parameters)
.doReturnStatus(200);

client.onGet("/foo").doReturnFormParams(Arrays.asList(
  new BasicNameValuePair("one", "1"),
  new BasicNameValuePair("one", "2")
));

PawelAdamski commented:
First of all thanks for the contribution. After the first review, your changes look good to me. Please give me some time to have a deeper look.

mangstadt commented:
Of course, let me know if you have any questions. Thanks for looking at it.

Sorry that the commit of my previous pull request was included in this pull request. I'm not too familiar with how pull requests work!

PawelAdamski commented on src/main/java/com/github/paweladamski/httpclientmock/RuleBuilder.java:
```diff
@@ -3,14 +3,17 @@
 
   private final List<Action> actions = new ArrayList<>();
   private final List<Condition> conditions = new ArrayList<>();
+  private UrlEncodedFormCondition formParametersCondition;
```
Code will simplify if you initialize `formParametersCondition` here. You can then remove all the `if formParametersCondition==null` statements.

When there are no conditions `UrlEncodedFormCondition#matches`  should remove true if and only if the request body have no form parameters.

mangstadt replied:
It still needs to check for when the user never calls `withFormParameters()` because he is not interested in checking them. In this use-case, the `matches` method should return true no matter how many form parameters the request has, because the user has not told it to check for any.

I can move this logic into the `UrlEncodedFormConditions` class so that the null checks can be removed.

mangstadt replied:
Fixed in d9ee25b7c17aa84fc00b36b6d5eeca68c22eae0b.

PawelAdamski commented on src/main/java/com/github/paweladamski/httpclientmock/condition/UrlEncodedFormCondition.java:
```diff
@@ -0,0 +1,77 @@
+  private final Map<String, Matcher<String>> expected = new HashMap<>();
+
+  @Override
+  public boolean matches(Request r) {
```
Please split this method into a method that will describe your logic step by step. 
My proposition:
```
public boolean matches(Request r) {
List<NameValuePair> actuallFormParams = findFormParameters(r);
return allDefinedParamsOccurredInRequest(actuallFormParams)
          && allParamsHaveMatchingValue(actuallFormParams);
```

Please see class UrlConditions how I have implemented methods:
`allDefinedParamsOccurredInRequest` and `allParamsHaveMatchingValue`

mangstadt replied:
Fixed in 67b2e5bbefcdc8287c547248e56f42fbedfb99ff.

PawelAdamski commented on src/main/java/com/github/paweladamski/httpclientmock/condition/UrlEncodedFormCondition.java:
```diff
@@ -0,0 +1,77 @@
+      return false;
+    }
+
+    Map<String, Matcher<String>> expectedCopy = new HashMap<>(expected);
```
Please see class UrlConditions how I have implemented methods:
allDefinedParamsOccurredInRequest and allParamsHaveMatchingValue

PawelAdamski commented on src/main/java/com/github/paweladamski/httpclientmock/condition/UrlEncodedFormCondition.java:
```diff
@@ -0,0 +1,77 @@
+ * @author Michael Angstadt
+ */
+public class UrlEncodedFormCondition implements Condition {
+  private final Map<String, Matcher<String>> expected = new HashMap<>();
```
Please use `MatcherMap` class.
 It supports multiple multiple conditions for a single parameter and has implemented `describe` method.

mangstadt replied:
Fixed in aa8deb5548dfde0ec6dc54be44189dad81327913.

PawelAdamski commented on src/main/java/com/github/paweladamski/httpclientmock/RuleBuilder.java:
```diff
@@ -3,14 +3,17 @@
+import java.util.Map;
 import org.hamcrest.Matcher;
 
 class RuleBuilder {
```
You forgot to add UrlEncdodedFormCondition to method `toRule`

mangstadt replied:
I add it using `RuleBuilder#addCondition`, so it is inside of the `conditions` list.

PawelAdamski commented on src/test/java/com/github/paweladamski/httpclientmock/HttpClientMockBuilderTest.java:
```diff
@@ -314,6 +321,97 @@ public void should_allow_different_host_then_default() throws IOException {
   }
+  
+  @Test
+  public void withFormParameter() throws IOException {
```
Please try to make this more compact, I think that we don't need some many requests to check if `verifing` work.

mangstadt replied:
I removed some of the code (see 06bbc4b5592745b41bead8038787dc70d90d1629). Is that better?

I suppose most of it is redundant, since the bulk of the functionality is in the `UrlEncodedFormCondition` class.

PawelAdamski replied:
Yes, now this test looks much better.

PawelAdamski commented on src/test/java/com/github/paweladamski/httpclientmock/HttpClientVerifyTest.java:
```diff
@@ -226,4 +233,103 @@ public void should_check_header() throws IOException {
   }
 
+  @Test
+  public void withFormParameter() throws IOException {
```
Please make this test more compact, I think you can test the same thing using a smaller number of test requests.

mangstadt replied:
I removed some of the code (see 06bbc4b5592745b41bead8038787dc70d90d1629). Is that better?

I suppose most of it is redundant, since the bulk of the functionality is in the `UrlEncodedFormCondition` class.

PawelAdamski commented on src/main/java/com/github/paweladamski/httpclientmock/HttpClientMockBuilder.java:
```diff
@@ -110,6 +137,17 @@ public HttpClientMockBuilder withBody(Matcher<String> matcher) {
+   * @param parameters the parameters
+   * @return condition builder
+   */
+  public HttpClientMockBuilder withBody(Map<String, Matcher<String>> parameters) {
```
I think name of this method is a bit confusing. If a method for setting a single form parameter is called `withFormParameter` than I would expect  that method for setting a collection of params would be `withFormParameters`

mangstadt replied:
Me too. Technically, form parameters are located inside of the request body. That's why I created the method. Changing the name to `withFormParameters` is OK, I think.

PawelAdamski commented on src/test/java/com/github/paweladamski/httpclientmock/HttpClientMockBuilderTest.java:
```diff
@@ -314,6 +321,97 @@ public void should_allow_different_host_then_default() throws IOException {
   }
+  
+  @Test
+  public void withFormParameter() throws IOException {
```
Does this test really work? I have doubts about it because you forgot to add `UrlEncodedFormCondition` to class `Rule` and this is necessary to actually check condition.

mangstadt replied:
The `UrlEncodedFormCondition` object was being added to the `RuleBuilder#conditions` list, which is why it works.

PawelAdamski commented on src/main/java/com/github/paweladamski/httpclientmock/condition/UrlEncodedFormCondition.java:
```diff
@@ -0,0 +1,77 @@
+ * Tests the request body for URL-encoded parameters.
+ * @author Michael Angstadt
+ */
+public class UrlEncodedFormCondition implements Condition {
```
Please add method `debug` and use it in `Rule#debug`, you can see example implementation in `UrlCondition` class.

mangstadt replied:
Fixed in 275ea2b73704e3ca0becfca11d924fa69ecf5d06.

PawelAdamski reviewed (changes requested):
One more I would like to thank you for your contribution. Every single line of code added to this project makes me very happy.

I have added a few comments. In case something is not clear, feel free to contact me.

mangstadt committed:
Initialize UrlEncodedFormCondition object in constructor

If the user never passes any expected form parameters into this object,
then it means the user isn't interested in checking for them. So, it
should return "true" no matter what the request looks like.

https://github.com/PawelAdamski/HttpClientMock/pull/20#discussion_r415086941

mangstadt committed:
Refactor for readability

https://github.com/PawelAdamski/HttpClientMock/pull/20#discussion_r415088578

mangstadt committed:
Use MatchersMap class

https://github.com/PawelAdamski/HttpClientMock/pull/20#discussion_r415089177

mangstadt committed:
Add debug method to UrlEncodedFormCondition class

mangstadt committed:
Fix indentation

mangstadt committed:
Removed redundant unit test code

https://github.com/PawelAdamski/HttpClientMock/pull/20#discussion_r415091895
https://github.com/PawelAdamski/HttpClientMock/pull/20#discussion_r415092756

PawelAdamski commented on src/main/java/com/github/paweladamski/httpclientmock/HttpClientVerifyBuilder.java:
```diff
@@ -84,6 +86,33 @@ public HttpClientVerifyBuilder withParameter(String name, Matcher<String> matche
+  /**
+   * Request body must contain the given URL-encoded form parameter (typically
+   * found in POST requests). Alternatively, parameters may be specified all at
+   * once using {@link #withBody(Map)}.
```
You forgot to change withBody int withFormParameters.

mangstadt replied:
Fixed in cb31d06f00a64792e996a5080e783986c22bbb16.

PawelAdamski commented on src/main/java/com/github/paweladamski/httpclientmock/HttpClientVerifyBuilder.java:
```diff
@@ -84,6 +86,33 @@ public HttpClientVerifyBuilder withParameter(String name, Matcher<String> matche
+  /**
+   * Request body must contain the given URL-encoded form parameter (typically
+   * found in POST requests). Alternatively, parameters may be specified all at
+   * once using {@link #withBody(Map)}.
```
You forgot to change withBody int withFormParameters.

mangstadt replied:
Fixed in cb31d06f00a64792e996a5080e783986c22bbb16.

PawelAdamski commented on src/main/java/com/github/paweladamski/httpclientmock/HttpClientMockBuilder.java:
```diff
@@ -110,6 +138,17 @@ public HttpClientMockBuilder withBody(Matcher<String> matcher) {
+   * @param parameters the parameters
+   * @return condition builder
+   */
+  public HttpClientMockBuilder withBody(MatchersMap<String, String> parameters) {
```
I thought you changed it to `withFormParameters'

mangstadt replied:
Fixed in cb31d06f00a64792e996a5080e783986c22bbb16.

PawelAdamski commented on src/test/java/com/github/paweladamski/httpclientmock/HttpClientMockBuilderTest.java:
```diff
@@ -314,6 +319,45 @@ public void should_allow_different_host_then_default() throws IOException {
     assertThat(login, hasContent("login"));
     assertThat(google, hasContent("google"));
   }
+  
```
I think there should be also a negative test. I mean we should also test the situation when `UrlEncodedFormCondition` fails.

mangstadt replied:
Added in f7d162730ad72a9549c7f2c7afeb45281589eade.

PawelAdamski commented on src/main/java/com/github/paweladamski/httpclientmock/condition/UrlEncodedFormCondition.java:
```diff
@@ -0,0 +1,131 @@
+ */
+public class UrlEncodedFormCondition implements Condition {
+  private final MatchersMap<String, String> expected = new MatchersMap<>();
+  private boolean enabled = false;
```
Do we really need this parameter?
I think it is unnecessary. When the `expected` map is empty method should return true only if request has zero form parameters.

Now it is actually not possible to say that I want to have a request without form parameters.

For URL parameters it already works in a similar way.

mangstadt replied:
Oh, I thought it would fail if the request body contained something other than form parameters. But it does not. Fixed in 78451eae763214a9bb05273999d1d4888e9cb554.

PawelAdamski commented on src/main/java/com/github/paweladamski/httpclientmock/condition/UrlEncodedFormCondition.java:
```diff
@@ -0,0 +1,131 @@
+  }
+  
+  @Override
+  public void debug(Request r, Debugger debugger) {
```
Nice, now it will be clear why condition failed.  :)

mangstadt committed:
Remove "enabled" field from UrlEncodedFormCondition

There is no need for this. If a request has a body that contains
something other than form parameters, the "URLEncodedUtils.parse" method
will not attempt to parse it and will return an empty list. It will only
parse the body if the request has a "application/x-www-form-urlencoded"
content type.

mangstadt committed:
Add test for when form parameters do not match

mangstadt committed:
Rename "withBody(MatchersMap)" method to "withFormParameters"

PawelAdamski commented on src/main/java/com/github/paweladamski/httpclientmock/condition/UrlEncodedFormCondition.java:
```diff
@@ -57,8 +53,18 @@ private boolean allParamsHaveMatchingValue(List<NameValuePair> actual) {
     try {
-      return URLEncodedUtils.parse(request.getEntity());
+      /*
+       * The method below returns an empty list if the Content-Type of the
```
Thanks, for the comments. It is always good to explain not obvious steps.

PawelAdamski reviewed (approved):
Now it looks good to me. Good job, thanks for the contribution.

Today I will push it to the Maven Central, usually, it takes one or two days until it will be available publicly.

mangstadt commented:
Sure thing!
